include/lo2s/perf/event.hpp

```diff
@@ -0,0 +1,291 @@
+    {
+        unit_ = unit;
+    }
+
+    void set_clock_attrs([[maybe_unused]] const bool& use_clockid,
+                         [[maybe_unused]] const clockid_t& clockid)
+    {
+#ifndef USE_HW_BREAKPOINT_COMPAT
+        attr_.use_clockid = use_clockid;
+        attr_.clockid = clockid;
+#endif
+    }
+
+    // When we poll on the fd given by perf_event_open, wakeup, when our buffer is 80% full
+    // Default behaviour is to wakeup on every event, which is horrible performance wise
+    void set_watermark(const size_t& mmap_pages)
+    {
+        attr_.watermark = 1;
+        attr_.wakeup_watermark = static_cast<uint32_t>(0.8 * mmap_pages * sysconf(_SC_PAGESIZE));
+    }
+
+    void set_exclude_kernel(const bool& exclude_kernel)
+    {
+        attr_.exclude_kernel = exclude_kernel;
+    }
+
+    void set_sample_period(const int& period);
+    void set_sample_freq(const uint64_t& freq);
+    void event_attr_update(std::uint64_t value, const std::string& format);
+
+    void parse_pmu_path(const std::string& ev_name);
+    void parse_cpus();
+    const std::set<Cpu>& supported_cpus() const;
+
+    bool is_valid() const;
+    bool event_is_openable();
+
+    bool is_available_in(ExecutionScope scope) const
+    {
+        // per-process should always work. the counter will just not count if the process is
+        // scheduled on a core that is not supprted by that counter
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    bool degrade_precision();
+
+    friend bool operator==(const Event& lhs, const Event& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator>(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+protected:
+    void update_availability();
+    void set_common_attrs(bool enable_on_exec);
+
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+
+    std::filesystem::path pmu_path_;
+    std::string pmu_name_;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on use_sampling_options() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public Event
+{
+public:
+    using Event::Event;
+    SysfsEvent(const std::string& ev_name, bool enable_on_exec = false);
+
+    void use_sampling_options(const bool& use_pebs, const bool& sampling, const bool& enable_cct);
+};
+
+/**
+ * Contains an opened instance of Event.
+ * Use any Event.open() method to construct an object
+ */
+class EventGuard
+{
+public:
+    EventGuard();
```

Comment:
There should not be default constructed EventGuard instances. the default constructor should be ` = delete`.